modules/bluefish/consumer/bluefish_consumer.cpp

```diff
@@ -254,56 +542,59 @@ struct bluefish_consumer : boost::noncopyable
+				else
+					A_memcpy(dest, frame.image_data().begin(), frame.image_data().size());
+			}
 			else
-				A_memcpy(reserved_frames_.front()->image_data(), frame.image_data().begin(), frame.image_data().size());
-		}
-		else
-			A_memset(reserved_frames_.front()->image_data(), 0, reserved_frames_.front()->image_size());
-								
+				A_memset(dest, 0, reserved_frames_.front()->image_size());
 
-		// Send and display
+			frame_timer_.restart();
 
-		if(embedded_audio_)
-		{
-			auto remapped_audio	= channel_remapper_.mix_and_rearrange(frame.audio_data());
-			auto frame_audio	= core::audio_32_to_24(remapped_audio);
-			encode_hanc(reinterpret_cast<BLUE_UINT32*>(reserved_frames_.front()->hanc_data()), 
-						frame_audio.data(), 
-						static_cast<int>(frame.audio_data().size()/channel_layout_.num_channels), 
-						static_cast<int>(channel_layout_.num_channels));
-								
-			blue_->system_buffer_write_async(const_cast<uint8_t*>(reserved_frames_.front()->image_data()), 
-											static_cast<unsigned long>(reserved_frames_.front()->image_size()), 
-											nullptr, 
-											BlueImage_HANC_DMABuffer(reserved_frames_.front()->id(), BLUE_DATA_IMAGE));
-
-			blue_->system_buffer_write_async(reserved_frames_.front()->hanc_data(),
-											static_cast<unsigned long>(reserved_frames_.front()->hanc_size()), 
-											nullptr,                 
-											BlueImage_HANC_DMABuffer(reserved_frames_.front()->id(), BLUE_DATA_HANC));
-
-			if(BLUE_FAIL(blue_->render_buffer_update(BlueBuffer_Image_HANC(reserved_frames_.front()->id()))))
-				CASPAR_LOG(warning) << print() << TEXT(" render_buffer_update failed.");
-		}
-		else
-		{
-			blue_->system_buffer_write_async(const_cast<uint8_t*>(reserved_frames_.front()->image_data()),
-											static_cast<unsigned long>(reserved_frames_.front()->image_size()), 
-											nullptr,                 
-											BlueImage_DMABuffer(reserved_frames_.front()->id(), BLUE_DATA_IMAGE));
-			
-			if(BLUE_FAIL(blue_->render_buffer_update(BlueBuffer_Image(reserved_frames_.front()->id()))))
-				CASPAR_LOG(warning) << print() << TEXT(" render_buffer_update failed.");
-		}
+			// remap, encode and copy hanc data
+			if (embedded_audio_)
+			{
+				auto remapped_audio = channel_remapper_.mix_and_rearrange(frame.audio_data());
+				auto frame_audio = core::audio_32_to_24(remapped_audio);
+				encode_hanc(reinterpret_cast<BLUE_UINT32*>(buf->hanc_data()),
+					frame_audio.data(),
+					static_cast<int>(frame.audio_data().size() / channel_layout_.num_channels),
+					static_cast<int>(channel_layout_.num_channels));
+			}
 
-		boost::range::rotate(reserved_frames_, std::begin(reserved_frames_)+1);
-		
+			live_frames_lock_.lock();
+			live_frames_.push(buf);
+			live_frames_lock_.unlock();
+
+			// start the thread if required.
+			if (dma_present_thread_ == 0)
+			{
+				end_dma_thread_ = false;
+				dma_present_thread_ = std::make_shared<std::thread>(&dma_present_thread_actual, this);
```

Comment:
Should be

```
dma_present_thread_ = std::make_shared<std::thread>([this]{dma_present_thread_actual();});
```

...is better because then you can change the `dma_present_thread_actual` function to be an instance member function instead.